records/2024-02-12.grouping-sms-codes.md

```diff
@@ -0,0 +1,123 @@
+to associate it only with templates of 2FA nature, i.e. the short code meant
+to send 2FA should be used to send 2FA SMS.
+
+At the current time, because the short code is sitting along long codes and all
+of these are in a default pool, that means that all codes are randomly used to
+send SMS, including the short codes. Hence the latter can be used to send
+SMS messages that are non-related with 2FA.
+
+To overcome this issue:
+
+- we need to isolate the short code and long codes within their own group that will be identified with proper usage,
+- we need a way to identify the templates with two-factor authentication.
+
+The ability to create a group of codes has extra benefits that we can leverage
+on top of solving the current issue:
+
+- the ability to have a formal random pool of long codes,
+- the ability to assign multiple dedicated long codes to one service,
+- the ability to repatriate the existing dedicated long codes from the us-west-1 region into the ca-central-1 and align better with our own policies of sending from Canada exclusively.
+
+### Overhaul of SMS sending
+
+This fits into our vision for a general overhaul of how we send SMS. The
+following areas along with broad tasks are envisioned at the time of this
+writing:
+
+```mermaid
+mindmap
+  root((short code))
+    Manage keywords
+    ::icon(fa-solid fa-key)
+        Identify replied service
+        Manage internal service block lists
+            Store block list per service
+            Report on the block list per service
+            Remove number from block list
+    Associate 2FA with templates
+    ::icon(fa-solid fa-link)
+        Provide usage on template form
+        ::icon(fa-brands fa-wpforms)
+        Identify existing 2FA template
+        ::icon(fa-solid fa-magnifying-glass)
+    Sending 2FA
+    ::icon(fa fa-comment-sms)
+        Update logic for original numbers
+        ::icon(fa-solid fa-code)
+            Fetch code with template
+            Fetch code with service
+            Fallback to default pool
+        Send SMS through Pinpoint
+        Create pool for short code
+    Repatriate dedicated codes to Canada
+    ::icon(fa-brands fa-canadian-maple-leaf)
+        Create pool for existing services with dedicated code
+        Auto pool creation on code assignment
+        Auto pool destruction on code removal
+```
+
+## Options
+
+Below are the different options that are considered to group SMS codes
+together.
+
+### Option 1: Leverage AWS Pinpoint pools
+
+Create a phone pool in AWS Pinpoint for regular long codes which act as a
+fallback. Create a pool containing the short code. Resurrect the Pinpoint
+implementation to send SMS and leverage the botocore Pinpoint API to send
+messages identified with a specific phone pool.
+
+#### Pros
+
+- Fast to implement
+
+#### Cons
+
+- Still cannot track which numbers that were used.
+- Have to use Pinpoint with a more complicated API. We have an old implementation that we can revive for this purpose.
```

Comment:
The old implementation was removed before [pools were introduced in 2022](https://github.com/aws/aws-sdk-go/blob/main/CHANGELOG.md#release-v14330-2022-03-31). The difference in sending with the new api and pools is a minorish change.

Reply:
Is that much different from the old implementation? just curious how much their API changed.

Reply:
Not really, which is good! Main differences really are
- there's a bit more to have set up - ie the pools and a "configuration set" that's used to tell it to send up the delivery receipt
- the receipt has differently named and organized fields, the the receipt processing task has to be slightly different.

(see the [terraform](https://github.com/cds-snc/notification-terraform/pull/1246/files) and [api](https://github.com/cds-snc/notification-api/pull/2152/files) WIP PRs for more details)